pkg/registry/flowcontrol/ensurer/strategy.go

```diff
@@ -58,112 +59,138 @@ type EnsureStrategy interface {
+// configurationObject is the relevant interfaces that each API object type implements
 type configurationObject interface {
 	metav1.Object
 	runtime.Object
 }
 
+// configurationObjectType adds the type constraint `comparable` and is thus
+// only usable as a type constraint.
+type configurationObjectType interface {
+	comparable
+	configurationObject
+}
+
+type objectOps[ObjectType configurationObjectType] struct {
+	client[ObjectType]
+	cache[ObjectType]
+	deepCopy    func(ObjectType) ObjectType
+	replaceSpec func(ObjectType, ObjectType) ObjectType
+	specEqual   func(expected, actual ObjectType) bool
+}
+
+func NewObjectOps[ObjectType configurationObjectType](client client[ObjectType], cache cache[ObjectType],
+	deepCopy func(ObjectType) ObjectType,
+	replaceSpec func(ObjectType, ObjectType) ObjectType,
+	specEqual func(expected, actual ObjectType) bool,
+) ObjectOps[ObjectType] {
+	return objectOps[ObjectType]{client: client, cache: cache,
+		deepCopy:    deepCopy,
+		replaceSpec: replaceSpec,
+		specEqual:   specEqual}
+}
+
+func (oo objectOps[ObjectType]) DeepCopy(obj ObjectType) ObjectType { return oo.deepCopy(obj) }
+
+func (oo objectOps[ObjectType]) ReplaceSpec(into, from ObjectType) ObjectType {
+	return oo.replaceSpec(into, from)
+}
+
+func (oo objectOps[ObjectType]) SpecEqual(expected, actual ObjectType) bool {
+	return oo.specEqual(expected, actual)
+}
+
 // NewSuggestedEnsureStrategy returns an EnsureStrategy for suggested config objects
-func NewSuggestedEnsureStrategy() EnsureStrategy {
-	return &strategy{
-		alwaysAutoUpdateSpecFn: func(_ wantAndHave) bool {
+func NewSuggestedEnsureStrategy[ObjectType configurationObjectType]() EnsureStrategy[ObjectType] {
+	return &strategy[ObjectType]{
+		alwaysAutoUpdateSpecFn: func(want, have ObjectType) bool {
 			return false
 		},
 		name: "suggested",
 	}
 }
 
 // NewMandatoryEnsureStrategy returns an EnsureStrategy for mandatory config objects
-func NewMandatoryEnsureStrategy() EnsureStrategy {
-	return &strategy{
-		alwaysAutoUpdateSpecFn: func(_ wantAndHave) bool {
+func NewMandatoryEnsureStrategy[ObjectType configurationObjectType]() EnsureStrategy[ObjectType] {
+	return &strategy[ObjectType]{
+		alwaysAutoUpdateSpecFn: func(want, have ObjectType) bool {
 			return true
 		},
 		name: "mandatory",
 	}
 }
 
 // auto-update strategy for the configuration objects
-type strategy struct {
-	alwaysAutoUpdateSpecFn func(wantAndHave) bool
+type strategy[ObjectType configurationObjectType] struct {
+	alwaysAutoUpdateSpecFn func(want, have ObjectType) bool
 	name                   string
 }
 
-func (s *strategy) Name() string {
+func (s *strategy[ObjectType]) Name() string {
 	return s.name
 }
 
-func (s *strategy) ShouldUpdate(wah wantAndHave) (updatable, bool, error) {
-	current := wah.getHave()
-
-	if current == nil {
-		return nil, false, nil
+func (s *strategy[ObjectType]) ShouldUpdate(objectOps objectLocalOps[ObjectType], current, bootstrap ObjectType) (ObjectType, bool, error) {
```

Comment:
Given it returns updated object, I suggest renaming this function.

`UpdateIfNeeded` maybe?

Reply:
Done, along with all the other copies of that name.